packages/core/src/conversation/ConversationService/ConversationService.ts

```diff
@@ -388,50 +424,173 @@ export class ConversationService extends TypedEventEmitter<Events> {
     groupId,
     conversationId,
     qualifiedUserIds,
-  }: RemoveUsersParams): Promise<Conversation> {
-    const clientsToRemove = await this.apiClient.api.user.postListClients({qualified_users: qualifiedUserIds});
+    shouldRetry = true,
+  }: RemoveUsersParams & {shouldRetry?: boolean}): Promise<Conversation> {
+    try {
+      const clientsToRemove = await this.apiClient.api.user.postListClients({qualified_users: qualifiedUserIds});
 
-    const fullyQualifiedClientIds = mapQualifiedUserClientIdsToFullyQualifiedClientIds(
-      clientsToRemove.qualified_user_map,
-    );
+      const fullyQualifiedClientIds = mapQualifiedUserClientIdsToFullyQualifiedClientIds(
+        clientsToRemove.qualified_user_map,
+      );
+
+      await this.mlsService.removeClientsFromConversation(groupId, fullyQualifiedClientIds);
 
-    await this.mlsService.removeClientsFromConversation(groupId, fullyQualifiedClientIds);
+      // key material gets updated after removing a user from the group, so we can reset last key update time value in the store
+      await this.mlsService.resetKeyMaterialRenewal(groupId);
+
+      return await this.getConversation(conversationId);
+    } catch (error) {
+      if (MLSService.isBrokenMLSConversationError(error)) {
+        if (!shouldRetry) {
+          this.logger.warn("Tried to remove users from MLS conversation but it's still broken after reset", error);
+          throw error;
+        }
+        this.logger.info(
+          "Tried to remove users from MLS conversation but it's broken, resetting the conversation",
+          error,
+        );
+        return this.handleBrokenMLSConversation(conversationId, newGroupId =>
+          this.removeUsersFromMLSConversation({
+            groupId: newGroupId,
+            conversationId,
+            qualifiedUserIds,
+            shouldRetry: false,
+          }),
+        );
+      }
 
-    //key material gets updated after removing a user from the group, so we can reset last key update time value in the store
-    await this.mlsService.resetKeyMaterialRenewal(groupId);
+      throw error;
+    }
+  }
 
-    return await this.getConversation(conversationId);
+  public async joinByExternalCommit(conversationId: QualifiedId, shouldRetry = true): Promise<void> {
+    try {
+      await this.mlsService.joinByExternalCommit(() => this.apiClient.api.conversation.getGroupInfo(conversationId));
+    } catch (error) {
+      if (MLSService.isBrokenMLSConversationError(error)) {
+        this.logger.info(
+          "Failed to join MLS conversation via external commit because it's broken, resetting the conversation",
+          error,
+        );
+        if (!shouldRetry) {
+          this.logger.warn("Tried to join MLS conversation but it's still broken after reset", error);
+          throw error;
+        }
+        return this.handleBrokenMLSConversation(conversationId, (_newGroupId: string) =>
+          this.joinByExternalCommit(conversationId, false),
```

Comment:
Isn't it redundant to join by external commit after you've "repaired" the broken conversation?